lib/internal/stream_base_commons.js

```diff
@@ -241,7 +251,20 @@ function setStreamTimeout(msecs, callback) {
       this.removeListener('timeout', callback);
     }
   } else {
-    this[kTimeout] = setUnrefTimeout(this._onTimeout.bind(this), msecs);
+    if (remaining > 0) {
+      this[kTimeout] = setUnrefTimeout(this._onTimeout.bind(this), remaining);
+      // Update timeout time after timer has been scheduled.
+      this[kTimeout]._idleTimeout = msecs;
+    } else if (elapsed > msecs) {
+      // TODO: setImmediate
+      this[kTimeout] = setUnrefTimeout(this._onTimeout.bind(this), 1);
+      // Update timeout time after timer has been scheduled.
+      this[kTimeout]._idleTimeout = msecs;
+    } else {
+      console.log("#3")
+      this[kTimeout] = setUnrefTimeout(this._onTimeout.bind(this), msecs);
+    }
+
     if (this[kSession]) this[kSession][kUpdateTimer]();
```

Comment:
This isn't pretty. I would very much like some help with a better approach. At least this illustrates what I'd like to achieve.